test/mainnet-fork/Uniswapv4.t.sol

```diff
@@ -0,0 +1,1971 @@
+
+        vm.prank(SPARK_PROXY);
+        mainnetController.setUniswapV4TickLimits(bytes32(0), 0, 1, 1); // lower must be less than upper
+
+        vm.prank(SPARK_PROXY);
+        vm.expectRevert("MC/invalid-ticks");
+        mainnetController.setUniswapV4TickLimits(bytes32(0), 0, 1, 0);
+
+        vm.prank(SPARK_PROXY);
+        mainnetController.setUniswapV4TickLimits(bytes32(0), 0, 0, 0); // maxTickSpacing can only be 0 if all 0
+    }
+
+    function test_setUniswapV4tickLimits() external {
+        vm.expectEmit(address(mainnetController));
+        emit MainnetController.UniswapV4TickLimitsSet(_POOL_ID, -60, 60, 20);
+
+        vm.record();
+
+        vm.prank(SPARK_PROXY);
+        mainnetController.setUniswapV4TickLimits(_POOL_ID, -60, 60, 20);
+
+        _assertReentrancyGuardWrittenToTwice();
+
+        ( int24 tickLowerMin, int24 tickUpperMax, uint24 maxTickSpacing ) = mainnetController.uniswapV4TickLimits(_POOL_ID);
+
+        assertEq(tickLowerMin,   -60);
+        assertEq(tickUpperMax,   60);
+        assertEq(maxTickSpacing, 20);
+    }
+
+    /**********************************************************************************************/
+    /*** Story Tests                                                                            ***/
+    /**********************************************************************************************/
+
+    /**
+     * @dev Story 1 is a round trip of liquidity minting, increase, decreasing, and closing/burning,
+     *      each 90 days apart, while an external account swaps tokens in and out of the pool.
+     *      - The relayer mints a position with 1,000,000 liquidity.
+     *      - The relayer increases the liquidity position by 50% (to 1,500,000 liquidity).
+     *      - The relayer decreases the liquidity position by 50% (to 750,000 liquidity).
+     *      - The relayer decreases the remaining liquidity position (to 0 liquidity).
+     */
+    function test_uniswapV4_story1() external {
+        // Setup the pool and the controller.
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setUniswapV4TickLimits(_POOL_ID, -60, 60, 20);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_DEPOSIT_LIMIT_KEY,  2_000_000e18, uint256(2_000_000e18) / 1 days);
+        rateLimits.setRateLimitData(_WITHDRAW_LIMIT_KEY, 2_000_000e18, uint256(2_000_000e18) / 1 days);
+        vm.stopPrank();
+
+        // 1. The relayer mints a position with 1,000,000 liquidity.
+        IncreasePositionResult memory increaseResult = _mintPosition(-10, 0, 1_000_000e6, type(uint160).max, type(uint160).max);
+
+        uint256 expectedDecrease = _to18From6Decimals(increaseResult.amount0Spent) + _to18From6Decimals(increaseResult.amount1Spent);
+        assertEq(2_000_000e18 - rateLimits.getCurrentRateLimit(_DEPOSIT_LIMIT_KEY), expectedDecrease);
+
+        // 2. 90 days elapse.
+        vm.warp(block.timestamp + 90 days);
+
```

Comment:
Assert underlying amounts for _getAmountsForLiquidity here before and after these swaps, they could just be happening outside of our range the whole time